### PR TITLE
Ensure nullable partitioning column is handled properly

### PR DESCRIPTION
**Pull request by fgwang7w**
```
== RELEASE NOTE ==

Hive changes
* Fix dynamic pruning failures for joining on null keys in hive partition
```

### CODE REVIEW

**Comment by fgwang7w** [comment]
resolve #15446

**Review by kewang1024 — CHANGES_REQUESTED** [review]
1. Can you merge those two commits into one commit?
2. Is it to solve a general issue or just for dynamic filtering? 
3. Can you add integration test as well? (Example like:TestHiveDistributedJoinQueriesWithDynamicFiltering, TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown )

**Comment by fgwang7w** [comment]
> 1. Can you merge those two commits into one commit?
> 2. Is it to solve a general issue or just for dynamic filtering?
> 3. Can you add integration test as well? (Example like:TestHiveDistributedJoinQueriesWithDynamicFiltering, TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown )

Thanks for reviewing the PR.
1. done
2. this issue happends when dynamic filtering is on, so the fix is to solve for DF only
3. TestHiveDistributedJoinQueriesWithDynamicFiltering.testJoinOnNullPartitioning is added for integration test.

**Comment by fgwang7w** [comment]
@kewang1024 could you please help review the code? it'd be ideal if the code can be merged in the upcoming release, thanks

**Comment by fgwang7w** [comment]
hi@highker could you please merge this fix? I'm running some benchmark testing with DF on and this is blocking in many tcs, thx

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 1]
```diff
@@ -524,8 +524,8 @@ private static boolean isValidPartitionType(Type type)
     public static NullableValue parsePartitionValue(String partitionName, String value, Type type, DateTimeZone timeZone)
     {
         verifyPartitionTypeSupported(partitionName, type);
-
-        boolean isNull = HIVE_DEFAULT_DYNAMIC_PARTITION.equals(value);
+        byte[] bytes = value.getBytes(UTF_8);
+        boolean isNull = HIVE_DEFAULT_DYNAMIC_PARTITION.equals(value) || isHiveNull(bytes);
```
Comment (kewang1024) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:528]: Let's merge these two lines

**Review by kewang1024 — APPROVED** [review]
1.For release note, it's not a general change, only hive change.
remove `== NO RELEASE NOTE ==`
2.Change the commit and release message to something like:
Fix dynamic pruning for null keys in hive partition

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 1 above)
Comment (fgwang7w) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: fixed, thanks for reviewing the PR

**Comment by kewang1024** [comment]
Also the release note please.
@highker can you help review this PR as well?

File: presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java [hunk 2]
```diff
@@ -116,6 +118,14 @@ public void testGetPartitionsSample()
         assertEquals(getPartitionsSample(ImmutableList.of(p1, p2, p3, p4, p5), 3), ImmutableList.of(p1, p5, p4));
     }
 
+    @Test
+    public void testNullablePartitionValue()
+    {
+        HivePartition part = partition("p1=name/p2=\\N");
+        NullableValue verify = new NullableValue(BIGINT, null);
+        assertTrue(part.getKeys().values().contains(verify));
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java:126]: `containsValue`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFiltering.java [hunk 3]
```diff
@@ -110,6 +110,22 @@ public void testJoinDynamicFilteringMultiJoin()
...
+        assertUpdate("INSERT INTO t4 VALUES(null, null), (2,2)", 2);
+
+        String query = "select * from t3, t4 where t3.c1=t4.c2";
+        Session session = Session.builder(getSession())
+                .setSystemProperty(ENABLE_DYNAMIC_FILTERING, "true")
+                .setSystemProperty(JOIN_DISTRIBUTION_TYPE, FeaturesConfig.JoinDistributionType.AUTOMATIC.name())
+                .setSystemProperty(JOIN_REORDERING_STRATEGY, FeaturesConfig.JoinReorderingStrategy.AUTOMATIC.name())
+                .build();
+        assertQuery(session, query, "SELECT null, 2, 2, 2");
+    }
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFiltering.java:128]: break a line after this

**Review by highker — APPROVED** [review]

File: presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java (hunk 2 above)
Comment (fgwang7w) [reply presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java:126]: Collection of NullableValue for partition key does not implement containsValue method, but contains method works fine

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFiltering.java (hunk 3 above)
Comment (fgwang7w) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFiltering.java:128]: is this the new coding style? I use airlift code style which removes the blank line.

File: presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java (hunk 2 above)
Comment (highker) [reply presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java:126]: What I meant is to change `values().contains()` to `containsValue`

Comment (fgwang7w) [reply presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestMetastoreHiveStatisticsProvider.java]: fixed

**Pull request merged**